File: backend/src/models/Booking.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const bookingSchema = new Schema({
  clientId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  professionalId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  skillId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Skill',
    required: true
  },
  date: {
    type: Date,
    required: true
  },
  timeSlot: {
    type: String,
    required: true
  },
  amount: {
    type: Number,
    required: true
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'in_progress', 'completed', 'cancelled'],
    default: 'pending'
  },
  completedAt: {
    type: Date,
    default: null
  },
  notes: {
    type: String,
    default: ''
  },
  location: {
    type: String,
    default: ''
  }
}, {
  timestamps: true
});

// Middleware per aggiornare completedAt quando lo status diventa 'completed'
bookingSchema.pre('findOneAndUpdate', function(next) {
  const update = this.getUpdate();
  if (update.status === 'completed' && !update.completedAt) {
    update.completedAt = new Date();
  }
  next();
});

// Metodi di utilità
bookingSchema.methods.isCompleted = function() {
  return this.status === 'completed';
};

bookingSchema.methods.isCancelled = function() {
  return this.status === 'cancelled';
};

bookingSchema.methods.isInProgress = function() {
  return this.status === 'in_progress';
};

bookingSchema.methods.getFormattedStatus = function() {
  const statusMap = {
    'pending': 'In attesa',
    'confirmed': 'Confermata',
    'in_progress': 'In corso',
    'completed': 'Completata',
    'cancelled': 'Cancellata'
  };
  return statusMap[this.status] || this.status;
};

// ✅ CORREZIONE: Evita la sovrascrittura del modello
module.exports = mongoose.models.Booking || mongoose.model('Booking', bookingSchema);

File: app/src/main/java/com/myzubster/data/model/Booking.kt
package com.myzubster.data.model

import com.google.gson.annotations.SerializedName

data class Booking(
    @SerializedName("id") val id: String,
    @SerializedName("skillId") val skillId: String,
    @SerializedName("clientId") val clientId: String,
    @SerializedName("professionalId") val professionalId: String,
    @SerializedName("status") val status: BookingStatus,
    @SerializedName("scheduledDate") val scheduledDate: String,
    @SerializedName("notes") val notes: String? = null,
    @SerializedName("price") val price: Double,
    @SerializedName("createdAt") val createdAt: String,
    @SerializedName("updatedAt") val updatedAt: String?
)

enum class BookingStatus {
    @SerializedName("pending") PENDING,
    @SerializedName("confirmed") CONFIRMED,
    @SerializedName("in_progress") IN_PROGRESS,
    @SerializedName("completed") COMPLETED,
    @SerializedName("cancelled") CANCELLED
}

data class CreateBookingRequest(
    @SerializedName("skillId") val skillId: String,
    @SerializedName("professionalId") val professionalId: String,
    @SerializedName("scheduledDate") val scheduledDate: String,
    @SerializedName("notes") val notes: String? = null
)

data class UpdateBookingRequest(
    @SerializedName("status") val status: BookingStatus? = null,
    @SerializedName("scheduledDate") val scheduledDate: String? = null,
    @SerializedName("notes") val notes: String? = null
)